src/Widgets.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bolt;
+
+use Bolt\Snippet\QueueProcessor;
+use Bolt\Snippet\Target;
+use Bolt\Snippet\Zone;
+use Bolt\Widget\BaseWidget;
+use Bolt\Widget\BoltHeaderWidget;
+use Bolt\Widget\CanonicalLinkWidget;
+use Bolt\Widget\NewsWidget;
+use Bolt\Widget\SnippetWidget;
+use Bolt\Widget\TwigAware;
+use Bolt\Widget\WeatherWidget;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Tightenco\Collect\Support\Collection;
+use Twig\Environment;
+
+class Widgets
+{
+    /** @var Collection */
+    private $queue;
+
+    /** @var Request */
+    private $request;
+
+    /** @var QueueProcessor */
+    private $queueProcessor;
+
+    /** @var Environment */
+    private $twig;
+
+    public function __construct(RequestStack $requestStack, QueueProcessor $queueProcessor, Environment $twig)
+    {
+        $this->queue = new Collection([]);
+        $this->request = $requestStack->getCurrentRequest();
+        $this->queueProcessor = $queueProcessor;
+        $this->twig = $twig;
+    }
+
+    public function registerWidget(BaseWidget $widget): void
+    {
+        $widget->setRequest($this->request);
+
+        if ($widget instanceof TwigAware) {
+            $widget->setTwig($this->twig);
+        }
+
+        $this->queue->push([
+            'priority' => $widget->getPriority(),
+            'target' => $widget->getTarget(),
+            'name' => $widget->getName(),
+            'zone' => $widget->getZone(),
+            'callback' => $widget,
+        ]);
+    }
+
+    public function renderWidgetByName(string $name): string
+    {
+        $widget = $this->queue->where('name', $name)->first();
+
+        if ($widget) {
+            return $widget['callback']();
```

Comment:
if instance of RequestAwareWidget then $widget($this->requestStack->getCurrentRequest())
(I assume object from queue will have __invoke)

Reply:
Done, but I kept it in `registerWidget`, for the reason mentioned above, _plus_ this way it'd need to be in _two_ methods.

Reply:
sending request to widget should be as lazy as possible because there can be more than one request in stack. Maybe push whole requestStack to it?